### PR TITLE
[PyTorch] Add sink attention support from cuDNN

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR adds sink attention support (fwd + bwd) to TE-PyTorch.
- `FusedAttention` backend for FP16/BF16 and BSHD/SBHD: cuDNN 9.13.1+ and cudnn-frontend 1.14.1
- `UnfusedDotProductAttention` backend for FP32/FP16/BF16 and BSHD/SBHD
- context parallel support for `cp_comm_type=a2a` with `FusedAttention`

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Added sink attention support via cuDNN
- Added `cp_comm_type` and `softmax_type` to `AttentionParams`
- Improved tensor allocation in `pytorch/csrc/extensions/attention.cpp` and tensor indexing in `Aux_CTX_Tensors` in `fused_attn_f16_arbitrary_seqlen.cu`
- Improved CP unit tests

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by jeremyyx** [comment]
Nice work! How could I find cudnn-frontend 1.14.1? I can only install 1.14.0. And I clone the latest cudnn-frontend, which still do not have "set_dsink_token"

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py [hunk 1]
```diff
@@ -533,6 +540,46 @@ def get_attention_backend(
...
+    # yes              | learnable    | FusedAttention
+    if softmax_type != "vanilla":
+        logger.debug("Disabling FlashAttention for softmax_type = %s", softmax_type)
+        use_flash_attention = False
+        if fp8 and fp8_meta["recipe"].fp8_dpa:
+            logger.debug("Disabling FusedAttention for softmax_type = %s in FP8", softmax_type)
+            use_fused_attention = False
+            logger.debug("Disabling UnfusedAttention for softmax_type = %s in FP8", softmax_type)
+            use_unfused_attention = False
+        if qkv_format == "thd":
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/utils.py:596]: Check if THD is supported.

**Comment by jeremyyx** [comment]
@cyanguwa  Thank you for the excellent work! I have merged the feature into my own Megatron fork, but I found that every attention backend reports “not supported” when qkv_format is set to 'thd'. Are there any work-arounds? At the moment, in packing mode the qkv format seems to be required to be 'thd'.

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py (hunk 1 above)
Comment (cyanguwa) [reply transformer_engine/pytorch/attention/dot_product_attention/utils.py:596]: We will not support THD at this time. Will try to add it in the next PR.

**Review by cuichenx — APPROVED** [review]
LGTM. Tested this branch with nemo training and convergence looks good.

**Comment by cyanguwa** [comment]
Tested with cuDNN 9.13.1 in pipeline 35266475. All looks good except for `cp_4_0` tests. Have reported them to cuDNN in bug 5522629. Based on @cuichenx's review and convergence testing, I'm merging the PR.

**Comment by cyanguwa** [comment]
> Nice work! How could I find cudnn-frontend 1.14.1? I can only install 1.14.0. And I clone the latest cudnn-frontend, which still do not have "set_dsink_token"

Sorry about the late reply. You probably have found my update to frontend in this PR to 1.14.1. The commit for FE now is [1a7b4b7](https://github.com/NVIDIA/cudnn-frontend/tree/1a7b4b78db44712fb9707d21cd2e3179f1fd88b8).

**Comment by cyanguwa** [comment]
> @cyanguwa Thank you for the excellent work! I have merged the feature into my own Megatron fork, but I found that every attention backend reports “not supported” when qkv_format is set to 'thd'. Are there any work-arounds? At the moment, in packing mode the qkv format seems to be required to be 'thd'.

We will add the support for 'thd' in our next PR. Due to release schedules, we didn't have time to push more changes in. We will support 'bshd' and 'sbhd' for now. Thanks!

File: tests/pytorch/attention/test_kv_cache.py [hunk 2]
```diff
@@ -469,7 +469,6 @@ def test_kv_cache(dtype, model, qkv_format, is_paged, backend, module, is_cuda_g
         config,
         qkv_dtype=dtype,
         qkv_layout=qkv_layout,
-        window_size=config.window_size,
```
Comment (ptrendx) [inline tests/pytorch/attention/test_kv_cache.py:472]: Why are we removing this?

File: tests/pytorch/utils.py [hunk 3]
```diff
@@ -147,12 +173,15 @@ def __init__(
         max_seqlen_kv: int = None,
         num_gqa_groups: int = None,
         head_dim_v: int = None,
+        softmax_type: str = "vanilla",
```
Comment (ptrendx) [inline tests/pytorch/utils.py:176]: maybe "base" instead?

File: tests/pytorch/attention/test_kv_cache.py (hunk 2 above)
Comment (cyanguwa) [reply tests/pytorch/attention/test_kv_cache.py:472]: It's part of the ModelConfig now (and being checked there). I tried to simplify `get_available_attention_backends` a bit.

File: transformer_engine/common/include/transformer_engine/fused_attn.h [hunk 4]
```diff
@@ -124,6 +124,18 @@ enum NVTE_Mask_Type {
...
+ *  \brief Attention softmax types
+ */
+enum NVTE_Softmax_Type {
+  /*! Vanilla softmax */
+  NVTE_VANILLA_SOFTMAX = 0,
+  /*! Off-by-one softmax */
+  NVTE_OFF_BY_ONE_SOFTMAX = 1,
+  /*! Learnable softmax */
+  NVTE_LEARNABLE_SOFTMAX = 2,
+};
```
Comment (ptrendx) [inline transformer_engine/common/include/transformer_engine/fused_attn.h:143]: Could you provide more information here on what those do? Maybe links to papers or networks that introduce those types of softmax?

Comment (cyanguwa) [reply transformer_engine/common/include/transformer_engine/fused_attn.h:143]: Added in [759b5ab](https://github.com/NVIDIA/TransformerEngine/pull/2148/commits/759b5ab71a0e70b34435c7317dbcf87fb653b1df).
